### PR TITLE
Simplified cChunkMap chunk storage

### PR DESCRIPTION
**Pull request by tigerw**
Funnily enough, on my system, on this particular world, std::map is at least as fast, sometimes faster than the array splitting approach we use:

Current:
```
     [19:55:03] Preparing spawn (world): 3.50% (14/400; 13.78 chunks / sec)
     [19:55:04] Preparing spawn (world): 13.00% (52/400; 37.25 chunks / sec)
     [19:55:04] Chunk generator performance: 53.68 ch / sec (108 ch total)
     [19:55:05] Preparing spawn (world): 26.75% (107/400; 53.45 chunks / sec)
     [19:55:06] Preparing spawn (world): 40.75% (163/400; 54.90 chunks / sec)
     [19:55:06] Chunk generator performance: 55.86 ch / sec (226 ch total)
     [19:55:07] Preparing spawn (world): 51.50% (206/400; 42.83 chunks / sec)
     [19:55:08] Preparing spawn (world): 64.50% (258/400; 51.74 chunks / sec)
     [19:55:08] Chunk generator performance: 54.90 ch / sec (333 ch total)
     [19:55:09] Preparing spawn (world): 77.75% (311/400; 52.68 chunks / sec)
     [19:55:10] Preparing spawn (world): 92.00% (368/400; 56.44 chunks / sec)
     [19:55:10] Chunk generator performance: 56.02 ch / sec (453 ch total)
     [19:55:11] Startup complete, took 10432ms!
Info [19:57:12] User cheesedummy10 authenticated with UUID 91c40b1eb8954561a6aa735b05632cc9
     [19:57:12] Player data file not found for cheesedummy10 (91c40b1eb8954561a6aa735b05632cc9, offline 18446de9338f3b28859a231072cb0568 (unused)), will be reset to defaults.
Info [19:57:12] Player cheesedummy10 has joined the game
     [19:57:15] Chunk generator performance: 40.03 ch / sec (48 ch total)
     [19:57:20] Chunk generator performance: 38.33 ch / sec (77 ch total)
     [19:57:22] Chunk generator performance: 41.86 ch / sec (168 ch total)
     [19:57:37] Chunk generator performance: 43.91 ch / sec (89 ch total)
     [19:57:40] Chunk generator performance: 46.36 ch / sec (93 ch total)
```

New:
```
     [19:40:25] Preparing spawn (world): 4.75% (19/400; 18.52 chunks / sec)
     [19:40:26] Preparing spawn (world): 20.50% (82/400; 61.95 chunks / sec)
     [19:40:27] Chunk generator performance: 68.30 ch / sec (137 ch total)
     [19:40:27] Preparing spawn (world): 37.00% (148/400; 64.83 chunks / sec)
     [19:40:28] Preparing spawn (world): 53.25% (213/400; 63.48 chunks / sec)
     [19:40:29] Chunk generator performance: 69.47 ch / sec (279 ch total)
     [19:40:30] Preparing spawn (world): 70.00% (280/400; 65.94 chunks / sec)
     [19:40:31] Chunk generator performance: 69.86 ch / sec (421 ch total)
     [19:40:31] Preparing spawn (world): 85.75% (343/400; 61.76 chunks / sec)
     [19:40:31] Startup complete, took 7274ms!
Info [19:40:36] User cheesedummy10 authenticated with UUID 91c40b1eb8954561a6aa735b05632cc9
     [19:40:36] Player data file not found for cheesedummy10 (91c40b1eb8954561a6aa735b05632cc9, offline 18446de9338f3b28859a231072cb0568 (unused)), will be reset to defaults.
Info [19:40:36] Player cheesedummy10 has joined the game
     [19:40:38] Chunk generator performance: 43.40 ch / sec (46 ch total)
     [19:40:44] Chunk generator performance: 51.85 ch / sec (105 ch total)
     [19:40:57] Chunk generator performance: 53.44 ch / sec (80 ch total)
     [19:41:00] Chunk generator performance: 52.39 ch / sec (81 ch total)
     [19:41:04] Chunk generator performance: 53.37 ch / sec (80 ch total)
```

### CODE REVIEW

**Comment by madmaxoft** [comment]
Silly GitHub won't show the diff for the exact file I'm interested in :P

**Comment by SphinxC0re** [comment]
use
```
meld --diff .
```

**Comment by tigerw** [comment]
Comments?

**Comment by SphinxC0re** [comment]
Seems okay.

**Comment by SphinxC0re** [comment]
Please rebase onto current master

**Comment by madmaxoft** [comment]
The core chunkmap functions handling the chunk getting and creation are in dire need of comments - what exactly they do, what they expect, how they're supposed to be called etc.

**Comment by madmaxoft** [comment]
Pity that I can't line-comment when GitHub doesn't show the differences

**Comment by tigerw** [comment]
How about I temporarily revert the m_CSChunkLayer -> m_CSChunks change so that the diff becomes viewable online?

**Comment by madmaxoft** [comment]
Try that, and rebase onto current master as well, please.

**Comment by SphinxC0re** [comment]
Any updates on this?

**Comment by tigerw** [comment]
There.

File: src/ChunkData.cpp [hunk 1]
```diff
@@ -44,26 +44,6 @@ cChunkData::cChunkData(cAllocationPool<cChunkData::sChunkSection> & a_Pool) :
...
-		{
-			return;
-		}
-	#endif
-	for (size_t i = 0; i < NumSections; i++)
-	{
-		Free(m_Sections[i]);
-		m_Sections[i] = nullptr;
-	}
-}
```
Comment (worktycho) [inline src/ChunkData.cpp]: Why? The free is absolutely necessary, unless you make m_Sections an array of unique_ptr, which should be done in a separate PR.

File: src/ChunkMap.h [hunk 2]
```diff
@@ -491,28 +431,27 @@ class cChunkMap
...
-	void RemoveLayer(cChunkLayer * a_Layer);
+	typedef std::list<cChunkStay *> cChunkStays;
 
 	cCriticalSection m_CSLayers;
-	cChunkLayerList  m_Layers;
-	cEvent           m_evtChunkValid;  // Set whenever any chunk becomes valid, via ChunkValidated()
+
+	/** A map of chunk coordinates to chunk pointers
+	Uses a map (as opposed to unordered_map) because sorted maps are apparently faster */
+	std::map<ChunkCoordinate, std::unique_ptr<cChunk>, ChunkCoordinate> m_Chunks;
```
Comment (worktycho) [inline src/ChunkMap.h]: How big was the world used to generate the statistics? Because std::map generally involves a tree, with O(log N growth)

File: src/ChunkData.cpp (hunk 1 above)
Comment (tigerw) [inline src/ChunkData.cpp]: I think I was trying stuff with ChunkData and not completely reverting it.

Why the __cplusplus macros by the way?

**Comment by worktycho** [comment]
I'm not convinced that spawn generation is a good benchmark. It happens rarely, involves a small world, and involves a much higher insert rate than normal use.

**Comment by worktycho** [comment]
Also, since we know the size of the spawn, could you try a version of the benchmark for unordered_map with the bucket size set to the size of the spawn zone? I'd like to know what the impact of rehashing is on the benchmark.

File: src/ChunkData.cpp (hunk 1 above)
Comment (worktycho) [inline src/ChunkData.cpp]: Because it was written before the project moved to requiring c++11. I used c++11 move constructors to ensure that I didn't accidentally copy the class, because thats expensive.

Basicly, everything in a __cplusplus < 201103L wrapper is obsolete, and needs to be removed when someone gets round to it. All the __cplusplus >= 201103L wrappers can be removed, because thats guarenteed now.

**Comment by tigerw** [comment]
The five "chunk generator performance:" messages you see were me moving around, so there's that.

Do you mean increasing the initial number of buckets (via constructor) to (PregenerateDistance * 2 + 1)², or changing the load factor?

**Comment by worktycho** [comment]
Increasing the initial bucket number.

**Comment by tigerw** [comment]
unordered map, bucket size (PregenerateDistance * 2 + 1)²
---
```
     [23:24:19] Preparing spawn (world): 8.00% (32/400; 31.75 chunks / sec)
     [23:24:20] Preparing spawn (world): 27.50% (110/400; 77.69 chunks / sec)
     [23:24:20] Chunk generator performance: 85.79 ch / sec (172 ch total)
     [23:24:21] Preparing spawn (world): 47.50% (190/400; 79.60 chunks / sec)
     [23:24:22] Preparing spawn (world): 67.25% (269/400; 78.14 chunks / sec)
     [23:24:22] Chunk generator performance: 86.37 ch / sec (346 ch total)
     [23:24:23] Preparing spawn (world): 86.50% (346/400; 77.00 chunks / sec)
     [23:24:24] Chunk generator performance: 85.54 ch / sec (482 ch total)
     [23:24:24] Startup complete, took 5937ms!
Info [23:24:30] User cheesedummy10 authenticated with UUID 91c40b1eb8954561a6aa735b05632cc9
     [23:24:30] Player data file not found for cheesedummy10 (91c40b1eb8954561a6aa735b05632cc9, offline 18446de9338f3b28859a231072cb0568 (unused)), will be reset to defaults.
Info [23:24:30] Player cheesedummy10 has joined the game
     [23:24:39] Chunk generator performance: 49.98 ch / sec (100 ch total)
```

map (as it stands now)
---
```
     [23:05:05] Preparing spawn (world): 10.00% (40/400; 39.92 chunks / sec)
     [23:05:06] Preparing spawn (world): 31.50% (126/400; 85.40 chunks / sec)
     [23:05:06] Chunk generator performance: 95.62 ch / sec (192 ch total)
     [23:05:07] Preparing spawn (world): 53.50% (214/400; 87.74 chunks / sec)
     [23:05:08] Preparing spawn (world): 75.00% (300/400; 84.73 chunks / sec)
     [23:05:08] Chunk generator performance: 94.82 ch / sec (381 ch total)
     [23:05:09] Preparing spawn (world): 95.50% (382/400; 80.71 chunks / sec)
     [23:05:09] Chunk generator performance: 94.11 ch / sec (478 ch total)
     [23:05:09] Startup complete, took 5767ms!
Info [23:05:21] User cheesedummy10 authenticated with UUID 91c40b1eb8954561a6aa735b05632cc9
     [23:05:21] Player data file not found for cheesedummy10 (91c40b1eb8954561a6aa735b05632cc9, offline 18446de9338f3b28859a231072cb0568 (unused)), will be reset to defaults.
Info [23:05:21] Player cheesedummy10 has joined the game
     [23:05:29] Chunk generator performance: 49.88 ch / sec (100 ch total)
     [23:05:30] Chunk generator performance: 47.92 ch / sec (146 ch total)
```

plain array (as it stands in master)
---
```
     [22:48:29] Preparing spawn (world): 5.25% (21/400; 20.57 chunks / sec)
     [22:48:30] Preparing spawn (world): 21.50% (86/400; 63.60 chunks / sec)
     [22:48:31] Chunk generator performance: 71.93 ch / sec (144 ch total)
     [22:48:31] Preparing spawn (world): 37.75% (151/400; 64.04 chunks / sec)
     [22:48:32] Preparing spawn (world): 55.00% (220/400; 68.18 chunks / sec)
     [22:48:33] Chunk generator performance: 71.80 ch / sec (288 ch total)
     [22:48:33] Preparing spawn (world): 71.00% (284/400; 64.00 chunks / sec)
     [22:48:34] Preparing spawn (world): 87.50% (350/400; 66.00 chunks / sec)
     [22:48:35] Chunk generator performance: 71.89 ch / sec (433 ch total)
     [22:48:35] Startup complete, took 7176ms!
Info [22:48:41] User cheesedummy10 authenticated with UUID 91c40b1eb8954561a6aa735b05632cc9
     [22:48:41] Player data file not found for cheesedummy10 (91c40b1eb8954561a6aa735b05632cc9, offline 18446de9338f3b28859a231072cb0568 (unused)), will be reset to defaults.
Info [22:48:41] Player cheesedummy10 has joined the game
     [22:48:50] Chunk generator performance: 45.75 ch / sec (92 ch total)
```

This is in no way rigorous.

**Comment by worktycho** [comment]
It might be worth looking at the results of my loadtester, to see how it performs on busy servers. But I'm convinced we need to swap to one of the standard collections, and switching from map to unordered_map later on is trivial.

**Comment by Seadragon91** [comment]
I have speedup of 7-11 chunks per seconds.

File: src/ChunkMap.cpp [hunk 3]
```diff
@@ -44,99 +44,35 @@ cChunkMap::cChunkMap(cWorld * a_World) :
...
-
-
-
-cChunkMap::~cChunkMap()
-{
-	cCSLock Lock(m_CSLayers);
-	while (!m_Layers.empty())
-	{
-		delete m_Layers.back();
-		m_Layers.pop_back();  // Must pop, because further chunk deletions query the chunkmap for entities and that would touch deleted data
```
Comment (tigerw) [inline src/ChunkMap.cpp]: Can anyone verify that this removal is correct? How does C++ go about automatically cleaning up data structures?

Comment (worktycho) [inline src/ChunkMap.cpp]: Its complicated. The default destructor won't take the lock whilst deleting the map, but I think its UB if anything else is touching the world whilst its being deleted, so that shouldn't matter.

**Comment by SphinxC0re** [comment]
Any updates on this one?

**Comment by tigerw** [comment]
I am being discriminated against for being busy! Nooooooo

**Comment by SphinxC0re** [comment]
Hmm... you didn't touch this since 2015 :laughing:

**Comment by tigerw** [comment]
uh,  :sweat_smile: very busy?

**Comment by tigerw** [comment]
Reopening for LogicParrot; I'm working on it.

File: src/Vector3.h [hunk 4]
```diff
@@ -393,11 +393,9 @@ class VectorHasher
...
 		// Guaranteed to have no hash collisions for any 128x128x128 area
 		size_t Hash = 0;
-		Hash ^= static_cast<size_t>(a_Vector.x);
-		Hash <<= 8;
-		Hash ^= static_cast<size_t>(a_Vector.y);
-		Hash <<= 8;
-		Hash ^= static_cast<size_t>(a_Vector.z);
+		Hash ^= a_Vector.x + 0x9e3779b9 + (Hash << 6) + (Hash >> 2);
+		Hash ^= a_Vector.y + 0x9e3779b9 + (Hash << 6) + (Hash >> 2);
+		Hash ^= a_Vector.z + 0x9e3779b9 + (Hash << 6) + (Hash >> 2);
```
Comment (LogicParrot) [inline src/Vector3.h]: Could you describe this hash function?

File: src/Vector3.h [hunk 5]
```diff
@@ -393,11 +393,9 @@ class VectorHasher
 	{
 		// Guaranteed to have no hash collisions for any 128x128x128 area
 		size_t Hash = 0;
-		Hash ^= static_cast<size_t>(a_Vector.x);
-		Hash <<= 8;
-		Hash ^= static_cast<size_t>(a_Vector.y);
```
Comment (LogicParrot) [inline src/Vector3.h]: I think there might be a bug in the old hash.

File: src/Vector3.h [hunk 6]
```diff
@@ -393,11 +393,9 @@ class VectorHasher
 	{
 		// Guaranteed to have no hash collisions for any 128x128x128 area
```
Comment (worktycho) [inline src/Vector3.h]: If you're going to remove this property, please also remove the comment.

File: src/Vector3.h (hunk 4 above)
Comment (worktycho) [inline src/Vector3.h]: This function appears to be a significant downgrade for our uses, as most of our vectors have similar magnitudes, so the lack of lower order collisions was benificial.

Comment (LogicParrot) [inline src/Vector3.h]: The lack of nearby collisions were perfect for the pathfinder because it scans locally. but I don't think the old hash function is good for the world as a whole; A worst case thought experiment: players can easily organize an "attack" by standing 128 chunks apart. If you can get enough of them to do this you are guaranteed to get so many collisions you're effectively now worse than a linked list. Note that the collisions will occur not just with the chunk where the players are standing, but in all loaded radius around them.

Comment (LogicParrot) [inline src/Vector3.h]: ...If we cannot find a reasonable hash, `std::map` may be a wiser choice.

Comment (worktycho) [inline src/Vector3.h]: The collisions are 256 chunks in z axis + 1 in y, or 256 in y + 1 in x, which restricts you to 256 collisions due to max height. But on 64 bit systems, we are only reducing the entropy by 8 bits, so how about:
```
size_t Hash = 0;
Hash ^= static_cast<size_t>(a_Vector.x);
Hash <<= sizeof(size_t) * CHAR_BIT / 2;
Hash ^= static_cast<size_t>(a_Vector.z);
Hash ^= static_cast<size_t>(a_Vector.y) & 0x0F<< CHAR_BIT *sizeof(size_t) - 4;
Hash ^= static_cast<size_t>(a_Vector.y) & static_cast<size_t>(~static_cast<What>(0x0F)) << CHAR_BIT *sizeof(size_t) / 2 -4;
```
It take x and z and uses them to fill the top and bottom half of the Hash. Then we take the bottom 4 bits of y and xor them into the top 4 bits of the hash, and any bits above the bottom four with the bits starting  4 bits below halfway. Guarantees no collision within any 2048x128x2048 region, or 2048x256x2048 region if y is capped between 0 and 255 on systems with a 32 bit size_t. On systems with a 64 bit size_t it has no collisions in a 134217728x128x134217728 / 134217728x256x134217728 area. I think 134217728x128x134217728 is sufficiently far apart to prevent players from causing such an attack without teleport permisions. Based on the wiki, assuming idea conditions, it would take nearly 39 days to achieve one set of collisions.

Comment (LogicParrot) [inline src/Vector3.h]: Wait, why is Y being taken into account at all, we're talking about *chunks*, they have no Y.

Comment (LogicParrot) [inline src/Vector3.h]: Which raises the question, why was the vector hash changed?

Comment (LogicParrot) [inline src/Vector3.h]: - The old hash was good for everything. And it's only used in the pathfinder.
 - I thought this was about chunk hashing and that confused me, since chunks have no Y.
 - Right now, I have no clue why this change was made, since vectors aren't used in the PR.

Comment (LogicParrot) [inline src/Vector3.h]: My thought experiment was about chunk hashing, by the way. Sorry for the confusion.

Comment (worktycho) [inline src/Vector3.h]: As a diagram on a 32 bit system:
```
x: 16 15 14 13 12 11 10  9  8  7  6  5  4  3  2  1  0
y:  3  2  1  0 21 20 19 18 17 16 15 14 13 12 11 10  9  8 7  6  5  4
z: 31 30 29 28 27 26 25 24 23 22 21 20 19 18 17 16 15 14 13 12 11 10 9 8 7 6 5 4 3 2 1 0
```
The each vertical column is the bits that are xored to get a result bit. The number is which bit from that component are we shifting where. For better entropy we could take to unused high bits of the x and y component, and xor them in at the bottom of the hash.

Comment (LogicParrot) [inline src/Vector3.h]: @worktycho The old hash is more than enough for its use case, we never put vectors in a hashtable other than in the pathfinder, and the 128x128x128 guarantee is enough for that. What exactly is the purpose of this proposed change or (@tigerw 's change), and is it related to the PR in any way?

Comment (worktycho) [inline src/Vector3.h]: I didn't see your comments whilst I was creating the diagram, and was suffering from exactly the same confusion. 

@tigerw you really need to split up your PRs.

Comment (tigerw) [inline src/Vector3.h]: This hash function, [obtained here](http://stackoverflow.com/questions/20511347/a-good-hash-function-for-a-vector), was provided to easily change between maps and unordered_maps with regards to chunk storage, especially since there is a lot of theorising about the speed of the latter.

The previous hash function was providing outputs which caused an unordered_map to place all entries in around 10%? of buckets, which degraded performance.

Since you're all talking about the hash function, I assume the rest of my code is perfect? ;)

Comment (tigerw) [inline src/Vector3.h]: > The collisions are 256 chunks in z axis + 1 in y, or 256 in y + 1 in x, which restricts you to 256 collisions due to max height. But on 64 bit systems, we are only reducing the entropy by 8 bits, so how about:

I think I may have been blinded by the sheer amount of CompSci in the following statement. I will attempt to digest this information.

Comment (LogicParrot) [inline src/Vector3.h]: @tigerw how does playing around with the VECTOR's hash function help you at all?

Comment (worktycho) [inline src/Vector3.h]: The chunk code wouldn't use Vector3 because the chunks are a 2D map. So this change seems to be useless, but if it is useful, it should be in a seperate PR.

Comment (tigerw) [inline src/Vector3.h]: > @tigerw how does playing around with the VECTOR's hash function help you at all?

That is, a very good question, come to think of it. I was using vectors to store chunk coordinates during testing.

**Comment by LogicParrot** [comment]
Could anyone commpare this to the current, post - #3153 master?

**Comment by LogicParrot** [comment]
Could anyone compare this to the current, post - #3153 master? (performance-wise)

**Comment by sphinxc0re** [comment]
So, of what pace are you working on this PR @tigerw?

**Comment by tigerw** [comment]
I shall revert the contested changes to Vector and make it merge able. I assume that the reviewers have no other objections to the code.

Workin' through mah prz, 1@a🕰
~ Tiger 2016

**Comment by tigerw** [comment]
Gonna merge this!

File: src/ChunkMap.h [hunk 7]
```diff
@@ -521,6 +459,7 @@ class cChunkMap
 
 	std::unique_ptr<cAllocationPool<cChunkData::sChunkSection> > m_Pool;
 
+	cChunkPtr ConstructChunk(int a_ChunkX, int a_ChunkZ);
```
Comment (madmaxoft) [inline src/ChunkMap.h]: There's still no documentation for this

File: src/ChunkMap.h [hunk 8]
```diff
@@ -492,27 +431,26 @@ class cChunkMap
...
+		int ChunkZ;
+	};
 
-	/** Returns the cChunkLayer object responsible for the specified chunk; creates it if not found. */
-	cChunkLayer * GetLayerForChunk (int a_ChunkX, int a_ChunkZ);
+	typedef std::list<cChunkStay *> cChunkStays;
 
-	/** Returns the specified cChunkLayer object; creates it if not found. */
-	cChunkLayer * GetLayer(int a_LayerX, int a_LayerZ);
+	cCriticalSection m_CSLayers;
```
Comment (madmaxoft) [inline src/ChunkMap.h]: `Layers` should be gone by this PR

**Comment by tigerw** [comment]
Now?

**Comment by tigerw** [comment]
Going once!

**Comment by tigerw** [comment]
Going twice!

File: src/ChunkMap.h [hunk 9]
```diff
@@ -492,27 +431,26 @@ class cChunkMap
...
 
-	typedef std::list<cChunkLayer *> cChunkLayerList;
-
-	typedef std::list<cChunkStay *> cChunkStays;
-
-	/** Finds the cChunkLayer object responsible for the specified chunk; returns nullptr if not found. Assumes m_CSLayers is locked. */
-	cChunkLayer * FindLayerForChunk(int a_ChunkX, int a_ChunkZ);
+	struct ChunkCoordinate
+	{
+		bool operator() (const ChunkCoordinate & a_Lhs, const ChunkCoordinate & a_Rhs) const
```
Comment (madmaxoft) [inline src/ChunkMap.h]: Wait, what? Why does a coordinate pair define call-operator, and why does the call operator compare two coordinates? This is very confusing at first look. Isn't it possible to make the comparison function a static one, and use it directly in the map's definition as the third template param?

Comment (tigerw) [inline src/ChunkMap.h]: I was just hoping to reuse the struct for both storage and comparison, so I wouldn't need to create an extra ChunkCoordinateComparer.

Comment (madmaxoft) [inline src/ChunkMap.h]: I figured that much, but it takes needless effort to figure, when it could be obvious. Make it an inner struct, `ChunkCoordinate::Comparer`.

**Comment by tigerw** [comment]
Merge?

**Comment by LogicParrot** [comment]
So, cChunkMap totally gone now?

**Comment by LogicParrot** [comment]
I meant cChunkLayer.

**Comment by tigerw** [comment]
\*poof\*
